File: python/xquant/orderbook.py
"""orderbook.py — XRPL DEX order book with microstructure metrics.

Mirrors include/xq/orderbook.hpp. Same XRPL caveats apply: the book is a CLOB
of Offers priced by 'quality' (TakerPays/TakerGets), XRP is integer drops,
offers settle at ledger close (~3-4s), and the native AMM (XLS-30) adds
liquidity a pure CLOB view misses.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from sortedcontainers import SortedDict  # pip install sortedcontainers


class Side(Enum):
    BID = "bid"
    ASK = "ask"


@dataclass
class FillResult:
    filled: float = 0.0
    avg_price: float = 0.0
    slippage_bps: float = 0.0
    complete: bool = False


class OrderBook:
    """Aggregated price-level book. Bids descending, asks ascending."""

    def __init__(self) -> None:
        # SortedDict keeps levels ordered for O(log n) top-of-book access,
        # the Python analogue of the C++ std::map choice (readability over
        # raw speed; the C++ side is the hot path).
        self._bids: SortedDict = SortedDict(lambda p: -p)  # highest first
        self._asks: SortedDict = SortedDict()              # lowest first

    def upsert(self, side: Side, price: float, qty: float) -> None:
        book = self._bids if side is Side.BID else self._asks
        if qty <= 0:
            book.pop(price, None)      # level cleared (offer removed)
        else:
            book[price] = qty

    def clear(self) -> None:
        self._bids.clear()
        self._asks.clear()

    # --- top of book --------------------------------------------------------
    def best_bid(self):
        return self._bids.peekitem(0) if self._bids else None   # (price, qty)

    def best_ask(self):
        return self._asks.peekitem(0) if self._asks else None

    def mid(self):
        if not self._bids or not self._asks:
            return None
        return (self._bids.peekitem(0)[0] + self._asks.peekitem(0)[0]) / 2

    def spread_bps(self):
        if not self._bids or not self._asks:
            return None
        b = self._bids.peekitem(0)[0]
        a = self._asks.peekitem(0)[0]
        return (a - b) / ((a + b) / 2) * 1e4

    def microprice(self):
        """Size-weighted fair value; leans toward the thin side (Stoikov)."""
        if not self._bids or not self._asks:
            return None
        b, qb = self._bids.peekitem(0)
        a, qa = self._asks.peekitem(0)
        tot = qb + qa
        if tot <= 0:
            return (a + b) / 2
        return (a * qb + b * qa) / tot   # weight each price by opposite size

    def imbalance(self, levels: int = 5):
        """Top-N order-book imbalance in [-1, 1]; >0 = bid-heavy."""
        if not self._bids or not self._asks:
            return None
        vb = sum(q for _, q in list(self._bids.items())[:levels])
        va = sum(q for _, q in list(self._asks.items())[:levels])
        tot = vb + va
        return (vb - va) / tot if tot > 0 else None

    def depth_within_bps(self, bps: float, side: Side):
        """Resting size within `bps` of mid = liquidity a taker can absorb."""
        m = self.mid()
        if m is None:
            return None
        acc = 0.0
        if side is Side.ASK:
            lim = m * (1 + bps / 1e4)
            for p, q in self._asks.items():
                if p > lim:
                    break
                acc += q
        else:
            lim = m * (1 - bps / 1e4)
            for p, q in self._bids.items():
                if p < lim:
                    break
                acc += q
        return acc

    def simulate_fill(self, take_side: Side, qty: float) -> FillResult:
        """Walk the book to fill `qty`; return VWAP + slippage vs arrival mid.
        take_side=ASK means we BUY (lift offers). A real market-impact model."""
        m = self.mid()
        book = self._asks if take_side is Side.ASK else self._bids
        remaining, notional = qty, 0.0
        for px, avail in book.items():
            take = min(remaining, avail)
            notional += take * px
            remaining -= take
            if remaining <= 1e-12:
                break
        r = FillResult()
        r.filled = qty - remaining
        r.complete = remaining <= 1e-12
        if r.filled > 0:
            r.avg_price = notional / r.filled
            if m:
                sgn = 1.0 if take_side is Side.ASK else -1.0
                r.slippage_bps = (r.avg_price - m) / m * 1e4 * sgn
        return r

    def take(self, take_side: Side, qty: float) -> FillResult:
        """Same walk as simulate_fill(), but actually CONSUMES the liquidity
        taken -- upserts each touched level down to its remaining size (or
        removes it if fully consumed). Needed for anything that fills more
        than once against the same book (e.g. a multi-child TWAP/VWAP
        execution -- see xquant/execution.py): calling simulate_fill()
        repeatedly against an unmutated book lets each call see the full
        original depth again, double-counting the same resting liquidity
        across child orders. simulate_fill() stays read-only on purpose (for
        "what would happen right now" queries that shouldn't have side
        effects); this is the one that actually happened."""
        m = self.mid()
        book = self._asks if take_side is Side.ASK else self._bids
        remaining, notional = qty, 0.0
        touched: list[tuple[float, float]] = []  # (price, remaining_qty_at_level)
        for px, avail in book.items():
            take = min(remaining, avail)
            notional += take * px
            remaining -= take
            touched.append((px, avail - take))
            if remaining <= 1e-12:
                break
        for px, remaining_qty in touched:
            self.upsert(take_side, px, remaining_qty)

        r = FillResult()
        r.filled = qty - remaining
        r.complete = remaining <= 1e-12
        if r.filled > 0:
            r.avg_price = notional / r.filled
            if m:
                sgn = 1.0 if take_side is Side.ASK else -1.0
                r.slippage_bps = (r.avg_price - m) / m * 1e4 * sgn
        return r

File: web/src/lib/orderbook.js
// orderbook.js -- aggregated price-level book + microstructure metrics.
// Ported from orderbook.py / orderbook.hpp / xrpl-desk.html's inline Map
// version; behavior kept identical across all three, this is just the one
// canonical JS copy.

export const Side = { BID: 'bid', ASK: 'ask' };

export class OrderBook {
  constructor() {
    this.bids = new Map(); // price -> qty
    this.asks = new Map();
  }

  upsert(side, price, qty) {
    const book = side === Side.BID ? this.bids : this.asks;
    if (qty <= 0) book.delete(price);
    else book.set(price, qty);
  }

  clear() {
    this.bids.clear();
    this.asks.clear();
  }

  sortedBids() {
    return [...this.bids.entries()].sort((a, b) => b[0] - a[0]); // highest first
  }

  sortedAsks() {
    return [...this.asks.entries()].sort((a, b) => a[0] - b[0]); // lowest first
  }

  bestBid() {
    const b = this.sortedBids();
    return b.length ? b[0] : null;
  }

  bestAsk() {
    const a = this.sortedAsks();
    return a.length ? a[0] : null;
  }

  mid() {
    const b = this.bestBid();
    const a = this.bestAsk();
    return b && a ? (b[0] + a[0]) / 2 : null;
  }

  spreadBps() {
    const b = this.bestBid();
    const a = this.bestAsk();
    if (!b || !a) return null;
    return ((a[0] - b[0]) / ((a[0] + b[0]) / 2)) * 1e4;
  }

  // Size-weighted fair value, leaning toward the thin side (Stoikov).
  microprice() {
    const b = this.bestBid();
    const a = this.bestAsk();
    if (!b || !a) return null;
    const [bp, bq] = b;
    const [ap, aq] = a;
    const tot = bq + aq;
    if (tot <= 0) return (ap + bp) / 2;
    return (ap * bq + bp * aq) / tot;
  }

  // Top-N order-book imbalance in [-1, 1]; >0 = bid-heavy.
  imbalance(levels = 5) {
    const b = this.sortedBids().slice(0, levels);
    const a = this.sortedAsks().slice(0, levels);
    if (!b.length || !a.length) return null;
    const vb = b.reduce((s, [, q]) => s + q, 0);
    const va = a.reduce((s, [, q]) => s + q, 0);
    const tot = vb + va;
    return tot > 0 ? (vb - va) / tot : null;
  }

  depthWithinBps(bps, side) {
    const m = this.mid();
    if (m == null) return null;
    let acc = 0;
    if (side === Side.ASK) {
      const lim = m * (1 + bps / 1e4);
      for (const [p, q] of this.sortedAsks()) {
        if (p > lim) break;
        acc += q;
      }
    } else {
      const lim = m * (1 - bps / 1e4);
      for (const [p, q] of this.sortedBids()) {
        if (p < lim) break;
        acc += q;
      }
    }
    return acc;
  }

  // Walk the book to fill `qty` on `takeSide` (ASK = we buy, lifting offers).
  // Returns VWAP + slippage vs arrival mid -- a real market-impact model.
  simulateFill(takeSide, qty) {
    const m = this.mid();
    const book = takeSide === Side.ASK ? this.sortedAsks() : this.sortedBids();
    let remaining = qty;
    let notional = 0;
    for (const [px, avail] of book) {
      const take = Math.min(remaining, avail);
      notional += take * px;
      remaining -= take;
      if (remaining <= 1e-12) break;
    }
    const filled = qty - remaining;
    const complete = remaining <= 1e-12;
    let avgPrice = 0;
    let slippageBps = 0;
    if (filled > 0) {
      avgPrice = notional / filled;
      if (m != null) {
        const sgn = takeSide === Side.ASK ? 1 : -1;
        slippageBps = ((avgPrice - m) / m) * 1e4 * sgn;
      }
    }
    return { filled, avgPrice, slippageBps, complete };
  }
}
